src/nupic/ntypes/Dimensions.hpp

```diff
@@ -29,352 +29,149 @@
-   */
-
-  /**
-   * Convert a Coordinate to a linear index (in lower-major order).
-   *
-   * @param coordinate
-   *        The coordinate to be converted
-   *
-   * @returns
-   *        The linear index corresponding to @a coordinate
-   */
-  size_t getIndex(const Coordinate &coordinate) const;
-
-  /**
-   * Convert a linear index (in lower-major order) to a Coordinate.
-   *
-   * @param index
-   *        The linear index to be converted
-   *
-   * @returns
-   *        The Coordinate corresponding to @a index
-   */
-  Coordinate getCoordinate(const size_t index) const;
-
-  /**
-   *
-   * @}
-   *
-   * @name Misc
-   *
-   * @{
-   *
-   */
-
-  /**
-   *
-   * Convert the Dimensions object to string representation.
-   *
-   * In most cases, we want a human-readable string, but for
-   * serialization we want only the actual dimension values
-   *
-   * @param humanReadable
-   *        The default is @c true, make the string human-readable,
-   *        set to @c false for serialization
-   *
-   * @returns
-   *        The string representation of the Dimensions object
-   */
-  std::string toString(bool humanReadable = true) const;
-
-
-  /**
-   * The equivalence operator.
-   *
-   * Two Dimensions objects will be considered equivalent, if any of the
-   * following satisfies:
-   *
-   * * They have the same number of dimensions and the same size for every
-   * dimension.
-   * * Both of them have the size of 1 for everything dimensions, despite of
-   * how many dimensions they have, i.e. isOnes() returns @c true for both
-   * of them. Some linking scenarios require us to treat [1] equivalent to [1 1]
-   * etc.
-   *
-   * @param dims2
-   *        The Dimensions object being compared
-   *
-   * @returns
-   *        Whether this Dimensions object is equivalent to @a dims2.
+   * * Invalid     - Some dimension is 0.
    *
+   * There is a function to check for each of these states.
    */
-  bool operator==(const Dimensions &dims2) const;
-
-  /**
-   * The in-equivalence operator, the opposite of operator==().
-   *
-   * @param dims2
-   *        The Dimensions object being compared
-   *
-   * @returns
-   *        Whether this Dimensions object is not equivalent to @a dims2.
-   */
-  bool operator!=(const Dimensions &dims2) const;
+  bool isUnspecified() const { return(size() == 0); }
+  bool isDontcare()    const { return(size() == 1 && at(0) == 0); }
+  bool isInvalid()     const { return(getCount() == 0); }
+  bool isSpecified()   const { return(!isInvalid()); }
+  static const int DONTCARE = 0;
 
-  /**
-   *
-   * @}
-   *
-   */
+  std::string toString(bool humanReadable = true) const {
+    if (isUnspecified()) return "[unspecified]";
+    if (isDontcare())    return "[dontcare]";
+    std::stringstream ss;
+    ss << "[";
+    for (size_t i = 0; i < size(); i++) {
+      if (i)  ss << "," <<at(i);
+      else   ss << at(i);
+    }
+    ss << "] ";
+		if (humanReadable && isInvalid()) ss << "(Invalid) ";
```

Comment:
please use same format as others, so `[invalid]`

Reply:
> if (humanReadable && isInvalid()) ss << "(Invalid) ";

Ok.  this is a little different from the others in that I show the value plus this (Invalid) text after it. 
```
 [4, 0] (invalid)  
```
 which was the original code.  But perhaps  `[4, 0] [invalid]`  would not look too bad.

Reply:
> different from the others in that I show the value plus this (Invalid) text after it 

if this is to be used, it should have the same format `[string invalid/dontcare/..] [value0, ...]` 
so this can be parsed or something. 

On the other hand, Occam's razor, do we really need this? Can the toString be removed?

Reply:
> do we really need this? Can the toString be removed?

When we switch to Cereal then it can replace toString().  For that reason I am not too concerned about getting it exactly right.  I need something here until we do the Cereal thing.

Reply:
my point was if you need to use toString somewhere? If not, just rm it (as Cereal..)